python/pyarrow/tests/test_flight.py

```diff
@@ -47,17 +48,52 @@ def do_get(self, context, ticket):
         return flight.RecordBatchStream(table)
 
 
+class MetadataFlightServer(flight.FlightServerBase):
+    """A Flight server that numbers incoming/outgoing data."""
+
+    def do_get(self, context, ticket):
+        data = [
+            pa.array([-10, -5, 0, 5, 10])
+        ]
+        table = pa.Table.from_arrays(data, names=['a'])
+        return flight.GeneratorStream(
+            table.schema,
+            MetadataFlightServer.number_batches(table))
+
+    def do_put(self, context, descriptor, reader, writer):
+        counter = 0
+        while True:
+            try:
+                batch, buf = reader.read_with_metadata()
+                assert buf is not None
+                client_counter, = struct.unpack('<i', buf.to_pybytes())
+                assert counter == client_counter
+                writer.write(struct.pack('<i', counter))
+                counter += 1
+            except StopIteration:
```

Comment:
Raising `StopIteration` doesn't sound great for a regular function or method. Perhaps return `None` instead?

Reply:
It is the same thing that `read_next_batch` does. Perhaps we should provide an iterator to wrap this method instead?

Reply:
Ah, right. Let's keep it like this then.